File: platform/rt-sandbox-ui/src/layout/rtLayoutMcProfile.ts
/**
 * Browser-side mirror of scripts/evaluation/rt_layout_mc_profile.py (dry-run preview only).
 * Keep behavior aligned with the Python translator; do not invoke Monte Carlo from the UI.
 */

import { sha256Hex16 } from "@/experiment/sha256Hex";

export const LAYOUT_SCHEMA_VERSION = "rt_layout_scenario_v1" as const;
export const PROFILE_SCHEMA_VERSION = "rt_layout_mc_profile_v1" as const;

const ENTITY_TYPE_SET = new Set<string>([
  "radar",
  "interceptor",
  "drone",
  "waypoint_marker",
]);

const REQUIRED_LAYOUT_KEYS = [
  "schema_version",
  "layout_id",
  "terrain_preset",
  "entities",
  "source",
] as const;

export type RtLayoutPose = {
  x: number;
  y: number;
  z: number;
  yaw_deg?: number;
};

export type RtLayoutEntity = {
  entity_type: string;
  pose: RtLayoutPose;
};

export type RtLayoutScenarioV1 = {
  schema_version: typeof LAYOUT_SCHEMA_VERSION;
  layout_id: string;
  terrain_preset: string;
  entities: RtLayoutEntity[];
  source: Record<string, unknown>;
  created_utc?: string;
  notes?: string;
};

export type LayoutValidation = {
  ok: boolean;
  issues: string[];
  warnings: string[];
};

export type McProfilePreview = {
  schema_version: typeof PROFILE_SCHEMA_VERSION;
  source_layout_id: string;
  geometry_id: string;
  scenario_suggestion: string;
  launch_args: string;
  launch_args_fields: Record<string, string>;
  entity_counts: Record<string, number>;
  warnings: string[];
  unsupported_fields: Array<{ field: string; reason: string }>;
  source: {
    artifact_schema: string;
    layout_source: unknown;
  };
  generated_utc: string;
};

function floatish(value: unknown): boolean {
  if (typeof value === "number" && Number.isFinite(value)) return true;
  if (typeof value === "string" && value.trim() !== "") {
    const n = Number(value);
    return Number.isFinite(n);
  }
  return false;
}

function normalizedPose(pose: Record<string, unknown>): RtLayoutPose {
  const out: RtLayoutPose = {
    x: Number(pose.x),
    y: Number(pose.y),
    z: Number(pose.z),
  };
  if (pose.yaw_deg !== undefined && pose.yaw_deg !== null) {
    out.yaw_deg = Number(pose.yaw_deg);
  }
  return out;
}

export function normalizedEntities(data: RtLayoutScenarioV1 | Record<string, unknown>): RtLayoutEntity[] {
  const entities: RtLayoutEntity[] = [];
  const raw = data.entities;
  if (!Array.isArray(raw)) return entities;
  for (const entity of raw) {
    if (!entity || typeof entity !== "object") continue;
    const row = entity as Record<string, unknown>;
    const pose = row.pose;
    if (!pose || typeof pose !== "object") continue;
    entities.push({
      entity_type: String(row.entity_type ?? ""),
      pose: normalizedPose(pose as Record<string, unknown>),
    });
  }
  return entities;
}

export function validateLayout(data: Record<string, unknown>): LayoutValidation {
  const issues: string[] = [];
  const warnings: string[] = [];

  for (const key of REQUIRED_LAYOUT_KEYS) {
    if (!(key in data)) {
      issues.push(`missing required field: ${key}`);
    }
  }

  if (data.schema_version !== LAYOUT_SCHEMA_VERSION) {
    issues.push(`schema_version must be '${LAYOUT_SCHEMA_VERSION}'`);
  }
  if (typeof data.layout_id !== "string" || !String(data.layout_id).trim()) {
    issues.push("layout_id must be a non-empty string");
  }
  if (typeof data.terrain_preset !== "string" || !String(data.terrain_preset).trim()) {
    issues.push("terrain_preset must be a non-empty string");
  }
  if (!data.source || typeof data.source !== "object") {
    issues.push("source must be an object");
  }

  const entities = data.entities;
  if (!Array.isArray(entities)) {
    issues.push("entities must be a list");
  } else if (entities.length === 0) {
    warnings.push("layout has no entities");
  } else {
    entities.forEach((entity, idx) => {
      if (!entity || typeof entity !== "object") {
        issues.push(`entities[${idx}] must be an object`);
        return;
      }
      const row = entity as Record<string, unknown>;
      const entityType = row.entity_type;
      if (!ENTITY_TYPE_SET.has(String(entityType))) {
        issues.push(`entities[${idx}].entity_type unsupported: ${String(entityType)}`);
      }
      const pose = row.pose;
      if (!pose || typeof pose !== "object") {
        issues.push(`entities[${idx}].pose must be an object`);
        return;
      }
      const poseRow = pose as Record<string, unknown>;
      for (const key of ["x", "y", "z"] as const) {
        if (!(key in poseRow)) {
          issues.push(`entities[${idx}].pose missing ${key}`);
        } else if (!floatish(poseRow[key])) {
          issues.push(`entities[${idx}].pose.${key} must be numeric`);
        }
      }
      if ("yaw_deg" in poseRow && !floatish(poseRow.yaw_deg)) {
        issues.push(`entities[${idx}].pose.yaw_deg must be numeric when present`);
      }
    });
  }

  return { ok: issues.length === 0, issues, warnings };
}

function stableStringify(value: unknown): string {
  if (value === null || typeof value !== "object") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map((item) => stableStringify(item)).join(",")}]`;
  }
  const obj = value as Record<string, unknown>;
  const keys = Object.keys(obj).sort();
  return `{${keys.map((key) => `${JSON.stringify(key)}:${stableStringify(obj[key])}`).join(",")}}`;
}

export function geometryFingerprint(data: RtLayoutScenarioV1 | Record<string, unknown>): string {
  const canonical = {
    schema_version: LAYOUT_SCHEMA_VERSION,
    terrain_preset: String(data.terrain_preset ?? ""),
    entities: normalizedEntities(data),
  };
  const digest = sha256Hex16(stableStringify(canonical));
  return `rt_layout:sha256:${digest}`;
}

/** Match Python f"{value:.6g}". */
export function fmtLaunchFloat(value: number): string {
  const s = Number(value).toPrecision(6);
  if (s.includes("e") || s.includes("E")) return s;
  return String(Number.parseFloat(s));
}

function countEntities(entities: RtLayoutEntity[]): Record<string, number> {
  const counts: Record<string, number> = {
    drone: 0,
    interceptor: 0,
    radar: 0,
    waypoint_marker: 0,
  };
  for (const entity of entities) {
    const key = entity.entity_type;
    if (key in counts) counts[key] += 1;
  }
  return counts;
}

export function translateLayoutToProfile(data: RtLayoutScenarioV1): McProfilePreview {
  const validation = validateLayout(data as unknown as Record<string, unknown>);
  if (!validation.ok) {
    throw new Error(validation.issues.join("; "));
  }

  const entities = normalizedEntities(data);
  const counts = countEntities(entities);
  const drones = entities.filter((e) => e.entity_type === "drone");
  const interceptors = entities.filter((e) => e.entity_type === "interceptor");
  const radars = entities.filter((e) => e.entity_type === "radar");
  const waypoints = entities.filter((e) => e.entity_type === "waypoint_marker");

  const warnings = [...validation.warnings];
  const unsupported_fields: Array<{ field: string; reason: string }> = [];
  const launch_fields: Record<string, string> = {};

  if (drones.length === 1) {
    const pose = drones[0].pose;
    launch_fields.target_start_x_m = fmtLaunchFloat(pose.x);
    launch_fields.target_start_y_m = fmtLaunchFloat(pose.y);
    launch_fields.target_start_z_m = fmtLaunchFloat(pose.z);
  } else if (drones.length === 0) {
    warnings.push("no drone entity: target_start_* launch args were not generated");
  } else {
    warnings.push(
      "multiple drone entities: MC target_start_* mapping supports exactly one drone today",
    );
    unsupported_fields.push({
      field: "entities[drone]",
      reason: "multiple drone layout is not launch-arg compatible today",
    });
  }

  if (interceptors.length === 3) {
    const coords: string[] = [];
    for (const entity of interceptors) {
      const pose = entity.pose;
      coords.push(fmtLaunchFloat(pose.x), fmtLaunchFloat(pose.y));
    }
    launch_fields.interceptor_ic_layout = `custom:${coords.join(",")}`;
  } else if (interceptors.length === 0) {
    warnings.push("no interceptor entities: interceptor_ic_layout was not generated");
  } else {
    warnings.push(
      "interceptor_ic_layout custom mapping requires exactly three interceptors today",
    );
    unsupported_fields.push({
      field: "entities[interceptor]",
      reason: `found ${interceptors.length} interceptor entities; expected 3`,
    });
  }

  for (const [entityType, rows] of [
    ["radar", radars],
    ["waypoint_marker", waypoints],
  ] as const) {
    if (rows.length > 0) {
      warnings.push(`${entityType} entities are retained as metadata only for MC preview`);
      unsupported_fields.push({
        field: `entities[${entityType}]`,
        reason: "no current Monte Carlo launch-arg mapping",
      });
    }
  }

  const scenario_suggestion = drones.length > 1 ? "multi" : "single";
  const launch_args = Object.entries(launch_fields)
    .map(([key, value]) => `${key}:=${value}`)
    .join(" ");

  return {
    schema_version: PROFILE_SCHEMA_VERSION,
    source_layout_id: data.layout_id,
    geometry_id: geometryFingerprint(data),
    scenario_suggestion,
    launch_args,
    launch_args_fields: launch_fields,
    entity_counts: counts,
    warnings,
    unsupported_fields,
    source: {
      artifact_schema: LAYOUT_SCHEMA_VERSION,
      layout_source: data.source,
    },
    generated_utc: new Date().toISOString().replace(/\.\d{3}Z$/, "Z"),
  };
}

File: scripts/evaluation/rt_layout_mc_profile.py
#!/usr/bin/env python3
"""Translate offline RT layout artifacts into Monte Carlo launch profile previews.

The translator is intentionally dry-run only: it does not launch Gazebo, mutate RT
sessions, write SA scenario packs, or promote replay artifacts.
"""

from __future__ import annotations

import argparse
import csv
import hashlib
import json
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

LAYOUT_SCHEMA_VERSION = "rt_layout_scenario_v1"
PROFILE_SCHEMA_VERSION = "rt_layout_mc_profile_v1"
ENTITY_TYPES = frozenset({"radar", "interceptor", "drone", "waypoint_marker"})
REQUIRED_LAYOUT_KEYS = ("schema_version", "layout_id", "terrain_preset", "entities", "source")
POSE_KEYS = ("x", "y", "z")


def _utc_now() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def _read_json_object(path: Path) -> dict[str, Any]:
    data = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(data, dict):
        raise ValueError(f"expected JSON object: {path}")
    return data


def _write_json(path: Path, data: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(data, indent=2, sort_keys=True) + "\n", encoding="utf-8")


def _floatish(value: Any) -> bool:
    try:
        float(value)
    except (TypeError, ValueError):
        return False
    return True


def validate_layout(data: dict[str, Any]) -> dict[str, Any]:
    """Validate an rt_layout_scenario_v1 object without touching runtime state."""
    issues: list[str] = []
    warnings: list[str] = []

    for key in REQUIRED_LAYOUT_KEYS:
        if key not in data:
            issues.append(f"missing required field: {key}")

    if data.get("schema_version") != LAYOUT_SCHEMA_VERSION:
        issues.append(f"schema_version must be {LAYOUT_SCHEMA_VERSION!r}")
    if not isinstance(data.get("layout_id"), str) or not str(data.get("layout_id")).strip():
        issues.append("layout_id must be a non-empty string")
    if not isinstance(data.get("terrain_preset"), str) or not str(data.get("terrain_preset")).strip():
        issues.append("terrain_preset must be a non-empty string")
    if not isinstance(data.get("source"), dict):
        issues.append("source must be an object")

    entities = data.get("entities")
    if not isinstance(entities, list):
        issues.append("entities must be a list")
        entities = []
    if len(entities) == 0:
        warnings.append("layout has no entities")

    for idx, entity in enumerate(entities):
        if not isinstance(entity, dict):
            issues.append(f"entities[{idx}] must be an object")
            continue
        entity_type = entity.get("entity_type")
        if entity_type not in ENTITY_TYPES:
            issues.append(f"entities[{idx}].entity_type unsupported: {entity_type!r}")
        pose = entity.get("pose")
        if not isinstance(pose, dict):
            issues.append(f"entities[{idx}].pose must be an object")
            continue
        for key in POSE_KEYS:
            if key not in pose:
                issues.append(f"entities[{idx}].pose missing {key}")
            elif not _floatish(pose[key]):
                issues.append(f"entities[{idx}].pose.{key} must be numeric")
        if "yaw_deg" in pose and not _floatish(pose["yaw_deg"]):
            issues.append(f"entities[{idx}].pose.yaw_deg must be numeric when present")

    return {"ok": len(issues) == 0, "issues": issues, "warnings": warnings}


def _normalized_pose(pose: dict[str, Any]) -> dict[str, float]:
    out = {key: float(pose[key]) for key in POSE_KEYS}
    if "yaw_deg" in pose:
        out["yaw_deg"] = float(pose["yaw_deg"])
    return out


def _normalized_entities(data: dict[str, Any]) -> list[dict[str, Any]]:
    entities: list[dict[str, Any]] = []
    for entity in data.get("entities") or []:
        if not isinstance(entity, dict):
            continue
        pose = entity.get("pose")
        if not isinstance(pose, dict):
            continue
        entities.append(
            {
                "entity_type": str(entity.get("entity_type") or ""),
                "pose": _normalized_pose(pose),
            }
        )
    return entities


def geometry_fingerprint(data: dict[str, Any]) -> str:
    """Return a stable geometry-only fingerprint for matched-seed MC pairing."""
    canonical = {
        "schema_version": LAYOUT_SCHEMA_VERSION,
        "terrain_preset": str(data.get("terrain_preset") or ""),
        "entities": _normalized_entities(data),
    }
    payload = json.dumps(canonical, sort_keys=True, separators=(",", ":")).encode("utf-8")
    digest = hashlib.sha256(payload).hexdigest()
    return f"rt_layout:sha256:{digest[:16]}"


def _fmt_float(value: float) -> str:
    return f"{value:.6g}"


def _count_entities(entities: list[dict[str, Any]]) -> dict[str, int]:
    counts = {key: 0 for key in sorted(ENTITY_TYPES)}
    for entity in entities:
        entity_type = str(entity.get("entity_type") or "")
        if entity_type in counts:
            counts[entity_type] += 1
    return counts


def translate_layout_to_profile(data: dict[str, Any]) -> dict[str, Any]:
    validation = validate_layout(data)
    if not validation["ok"]:
        raise ValueError("; ".join(validation["issues"]))

    entities = _normalized_entities(data)
    counts = _count_entities(entities)
    drones = [e for e in entities if e["entity_type"] == "drone"]
    interceptors = [e for e in entities if e["entity_type"] == "interceptor"]
    radars = [e for e in entities if e["entity_type"] == "radar"]
    waypoints = [e for e in entities if e["entity_type"] == "waypoint_marker"]

    warnings = list(validation["warnings"])
    unsupported_fields: list[dict[str, str]] = []
    launch_fields: dict[str, str] = {}

    if len(drones) == 1:
        pose = drones[0]["pose"]
        launch_fields["target_start_x_m"] = _fmt_float(pose["x"])
        launch_fields["target_start_y_m"] = _fmt_float(pose["y"])
        launch_fields["target_start_z_m"] = _fmt_float(pose["z"])
    elif len(drones) == 0:
        warnings.append("no drone entity: target_start_* launch args were not generated")
    else:
        warnings.append("multiple drone entities: MC target_start_* mapping supports exactly one drone today")
        unsupported_fields.append(
            {
                "field": "entities[drone]",
                "reason": "multiple drone layout is not launch-arg compatible today",
            }
        )

    if len(interceptors) == 3:
        coords: list[str] = []
        for entity in interceptors:
            pose = entity["pose"]
            coords.extend([_fmt_float(pose["x"]), _fmt_float(pose["y"])])
        launch_fields["interceptor_ic_layout"] = "custom:" + ",".join(coords)
    elif len(interceptors) == 0:
        warnings.append("no interceptor entities: interceptor_ic_layout was not generated")
    else:
        warnings.append("interceptor_ic_layout custom mapping requires exactly three interceptors today")
        unsupported_fields.append(
            {
                "field": "entities[interceptor]",
                "reason": f"found {len(interceptors)} interceptor entities; expected 3",
            }
        )

    for entity_type, rows in (("radar", radars), ("waypoint_marker", waypoints)):
        if rows:
            warnings.append(f"{entity_type} entities are retained as metadata only for MC preview")
            unsupported_fields.append(
                {
                    "field": f"entities[{entity_type}]",
                    "reason": "no current Monte Carlo launch-arg mapping",
                }
            )

    scenario_suggestion = "multi" if len(drones) > 1 else "single"
    launch_args = " ".join(f"{key}:={value}" for key, value in launch_fields.items())

    return {
        "schema_version": PROFILE_SCHEMA_VERSION,
        "source_layout_id": str(data["layout_id"]),
        "geometry_id": geometry_fingerprint(data),
        "scenario_suggestion": scenario_suggestion,
        "launch_args": launch_args,
        "launch_args_fields": launch_fields,
        "entity_counts": counts,
        "warnings": warnings,
        "unsupported_fields": unsupported_fields,
        "source": {
            "artifact_schema": LAYOUT_SCHEMA_VERSION,
            "layout_source": data.get("source"),
        },
        "generated_utc": _utc_now(),
    }


def profile_csv_row(profile: dict[str, Any]) -> dict[str, str]:
    return {
        "source_layout_id": str(profile.get("source_layout_id") or ""),
        "geometry_id": str(profile.get("geometry_id") or ""),
        "scenario_suggestion": str(profile.get("scenario_suggestion") or ""),
        "launch_args": str(profile.get("launch_args") or ""),
        "warning_count": str(len(profile.get("warnings") or [])),
        "unsupported_count": str(len(profile.get("unsupported_fields") or [])),
        "warnings": " | ".join(str(w) for w in profile.get("warnings") or []),
    }


def write_profile_csv(path: Path, profile: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    fields = [
        "source_layout_id",
        "geometry_id",
        "scenario_suggestion",
        "launch_args",
        "warning_count",
        "unsupported_count",
        "warnings",
    ]
    with path.open("w", encoding="utf-8", newline="") as handle:
        writer = csv.DictWriter(handle, fieldnames=fields)
        writer.writeheader()
        writer.writerow(profile_csv_row(profile))


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Preview MC launch args from an offline rt_layout_scenario_v1 artifact."
    )
    parser.add_argument("layout_json", type=Path)
    parser.add_argument("--out-json", type=Path, default=None)
    parser.add_argument("--out-csv", type=Path, default=None)
    args = parser.parse_args()

    try:
        layout = _read_json_object(args.layout_json)
        profile = translate_layout_to_profile(layout)
    except (OSError, json.JSONDecodeError, ValueError) as exc:
        print(f"rt_layout_mc_profile: {exc}", file=sys.stderr)
        return 2

    if args.out_json:
        _write_json(args.out_json, profile)
    if args.out_csv:
        write_profile_csv(args.out_csv, profile)
    if not args.out_json:
        print(json.dumps(profile, indent=2, sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
